File: Data Structures & Algorithms/valid-binary-search-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return isValid(root, INT_MIN, INT_MAX);
    }

    bool isValid(TreeNode* node, int left, int right){
        if(!node) return 1;
        int curr = node->val;
        
        if(!(left< curr && right> curr)) return 0;

        return isValid(node->left, left, node->val) && isValid(node->right, node->val, right);
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int goodNodes(TreeNode root) {
        int[] ans = new int[1];
        int maxi = Integer.MIN_VALUE;
        
        dfs(root, ans, maxi);
        return ans[0];
    }

    void dfs(TreeNode node, int[] ans, int maxi) {
        if(node == null) {
            return;
        }

        if(node.val >= maxi) 
            ans[0]++;
        
        maxi = Math.max(node.val, maxi);
        dfs(node.left, ans, maxi);
        dfs(node.right, ans, maxi);
    }
}
